### PR TITLE
Ps1 git discovery

### PR DESCRIPTION
**Pull request by daxgames**
## New PR, replacing #2381 because it works better and is less complicated.

I also have a PR ready for the `cmder::Cmder` sessions that follow the same logic described below.

This change affects Powershell sessions.

- Provides more consistent Powershell path with `cmd::cmder` sessions when using Cmder embedded Git.
- Allows a user to configure how Cmder modifies the path if a local Git is installed by adjusting the path position of the user Git folders.  See below.

See #2373 for the full proposal.

- Find user installed Git on Path
  - If found
    - if newer than Cmder embedded Git
      - Use it's existing Path config and completely ignore embedded Git.
    - Else if Cmder embedded Git exists and is newer
      - Match User installed Git path config using Cmder embedded Git folders.
  - Else if Cmder embedded Git exists 
    - Add Cmder embedded Git folders to the path.
      - `$env:cmder_root\vendor\git-for-windows\cmd;$env:path`
      - `$env:path;$env:cmder_root\vendor\git-for-windows\usr\bin`
      - `$env:path;$env:cmder_root\vendor\git-for-windows\mingw64\bin`

### CODE REVIEW

**Comment by daxgames** [comment]
@cmderdev/trusted-contributors - Can you please look at this and provide feedback.  This works but is a big change and I would like opinions keeping in mind I plan to make the `cmd::Cmder` sessions work the same way after this is merged.

You can install a newer/older version of Git externally, choose your path config during Git install.

- If the Git you install externally is newer than Cmder embedded Git:
  - Cmder Powershell sessions will use that config.
- If the Git you install is older than the Cmder embedded Git
   - Cmder Powershell sessions will match that config but use the embedded Git.

**Comment by daxgames** [comment]
Also this does not add any time to the Powershell session start time.

File: vendor/psmodules/Cmder.ps1
Comment (Jackbennett) [inline vendor/psmodules/Cmder.ps1:13]: With all the other `write-host` calls commented out `pause` here isn't going to have any context for the user.

**Review by Jackbennett — COMMENTED** [review]
Thought I'd try the review button, may have been a mistake not sure what I'm doing with github.

But take these as suggestions, there's a lot of excessive work with strings in here that powershell doesn't make you do. pipe variables to `fl *` to see what properties might be available.

File: vendor/profile.ps1
Comment (daxgames) [reply vendor/profile.ps1:46]: This happens before adding Cmder embedded Git(if present) to the path statement.   I see your point, adding the `-all`.

File: vendor/profile.ps1
Comment (daxgames) [reply vendor/profile.ps1:49]: See: https://github.com/cmderdev/cmder/issues/2373#issuecomment-682500553 for info on Git as a shim

Appreciate the more Powershell way of doing this.  I basically opened the cmd scripts that do this and rewrote in Powershell without trying to figure out the 'right' way to do it.  Will look at changing it.

File: vendor/psmodules/Cmder.ps1
Comment (daxgames) [reply vendor/psmodules/Cmder.ps1:72]: Prepending works if you are just prepending but this is truly reconfiguring the beginning and the middle of the path.

This does not just prepend the path where `git.exe` is but also the other two folders, `%GIT_INSTALL_ROOT%\usr\bin` and `%GIT_INSTALL_ROOT%\mingw64\bin`, the external Git install may or may not have added to the path either before or after the `c:\Windows;c:\Windows\system32` paths depending on what the user selected during install of Git for Windows.

I guess I could be less sneaky and replace `C:\Program Files\Git\usr\bin` with `%CMDER_ROOTvendor\git-for-windows\usr\bin;C:\Program Files\Git\usr\bin` and `%CMDER_ROOT%\vendor\git-for-windows\mingw64\bin;C:\Program Files\Git\mingw64\bin` leaving references to their original config in place instead of completely replacing it.

File: vendor/profile.ps1
Comment (daxgames) [reply vendor/profile.ps1:56]: Will update

File: vendor/psmodules/Cmder.ps1
Comment (daxgames) [reply vendor/psmodules/Cmder.ps1:25]: Could definitely be cleaner.

File: vendor/psmodules/Cmder.ps1
Comment (daxgames) [reply vendor/psmodules/Cmder.ps1:13]: I have no idea why that is there probably just an oversight while debugging.

File: vendor/psmodules/Cmder.ps1
Comment (Jackbennett) [reply vendor/psmodules/Cmder.ps1:72]: Ah ok, you might find it more pleasant to work with an array instead of strings;
```powershell
# array from .split cannot have size modified. Cast into an arrayList,
[System.Collections.ArrayList]$modifiedPath = $env:Path.split(';')`

$modifiedPath.removeAt(<index>) #remove specific index or...
$modifiedPath.remove('<string to match>') # remove specific string from collection

$env:path = $modifiedPath -join ';'
```

**Comment by MartiUK** [comment]
Happy to review this @daxgames, can you make sure you've looked at what @Jackbennett has already commented on?

**Comment by daxgames** [comment]
I will.

File: vendor/profile.ps1
Comment (daxgames) [reply vendor/profile.ps1:46]: changed to do `get-command -all`

File: vendor/profile.ps1
Comment (daxgames) [reply vendor/profile.ps1:49]: Using `$git.version`

@Jackbennett Not sure I understand what you mean by

> $gitdir might be better off for you down the script as $gitItem = get-item $git.path

File: vendor/profile.ps1
Comment (daxgames) [reply vendor/profile.ps1:52]: Using `[version]"2.28" -lt $git.Version`

File: vendor/profile.ps1
Comment (daxgames) [reply vendor/profile.ps1]: @Jackbennett this is not doing what you thought it was doing.  It is actually stripping off `/cmd` from the end of the string.  I have updated the code to make that more clear.

File: vendor/psmodules/Cmder.ps1
Comment (daxgames) [reply vendor/psmodules/Cmder.ps1]: fixed

**Comment by daxgames** [comment]
@MartiUK I have addressed all of @Jackbennett comments and responded to each.

File: vendor/psmodules/Cmder.ps1
Comment (daxgames) [reply vendor/psmodules/Cmder.ps1:71]: Did not look at this.  This was not easy to get to work and don't really have the time to completely rewrite it using arrays.

Comment (daxgames) [reply vendor/psmodules/Cmder.ps1:71]: If we don't want to merge because this is less than optimal I can see if I can carve out some time to look at it more closely.

**Comment by Jackbennett** [comment]
I've been spoilt I think as a gitlab admin elsewhere I can't add commits onto your pull request.

Refactored to simplify what we need to change to use the right git version https://github.com/cmderdev/cmder/commits/pr/2438

As None of these link to a portable git, they all use the OS configs. I don't think we need to try this hard to help a user that can't install git into a location they want to use it from? I'm not clear what problem is actually being solved. I propose to just find the latest git version and use that, adding some paths we know to look in for git.

I should have checked if gitshim still works but it's 2am. If that's no good I not blocking the PR, it just seems to do more than needs doing. non-admins can symlink in windows now too so I don't know why it exists https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/

**Comment by daxgames** [comment]
@MartiUK I put some finishing touches on this this AM and as far as I am concerned it is ready and fully implements the proposed functionality in the description of this PR.  If you would like to review for functionality please do so we can merge all PRs and cut a release.

There is a lot of good stuff to be released in `master` and I would like to get this change in before we do.

**Review by MartiUK — APPROVED** [review]
LGTM :shipit:

**Pull request merged**
